Refactoring of TagsDialog

TarekkMA opened the pull request:
## Pull Request template

- [x] Cleanup & Refactoring
- [x] Unit tests
    - [x] Still Have to test `TagsList` class
    - [x] UniqueArrayList
- [ ] ~Add multi note tag edit~ (will be done on another PR)

## Fixes
Related to #4911 
Related to #8327

## Approach
I decided to clean up and refactor `TagDialog` first and then move on implementing multi note tag edit next.

I implemented a new collection class `UniqueArrayList`, to be able to keep a list with a unique items, while keeping order and to have the ability to sort them

## How Has This Been Tested?

* Manually and using some unit tests

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:
```diff
@@ -0,0 +1,189 @@
+ * <p>
+ * The implementation of this class uses {@link ArrayList} as a base, to enable easy sorting and indexing.
+ */
+public class UniqueArrayList<E> extends ArrayList<E> {
```
Implementation can be improved.

Maybe keeping an internal set to do contain checks.

TarekkMA commented:
**I will go through implementing the following when the review on the refactor is okay**

My approach for adding multi note tagging is as follows:
* Callers of the dialog will have to send:
   * Checked Tags (sum of all notes checked tags)
   * Unchecked Tags (sum of all notes unchecked tags)
* `TagsList` will host another set called `mIndeterminate` which will be the intersection between the checked and the unchecked
* `TagsList` will have a method called `isIndeterminate`
* `TagsList# check()/uncheck()` will move the tag if found from  `mIndeterminate` to `mChecked` or `mAllTags`
* `TagsDialogListener.onPositive` will pass extra list of all indeterminate tags
```java
    public interface TagsDialogListener {
        void onPositive(List<String> selected, List<String> indeterminate , int option);
    }
```
* when listeners receive the call if tag is found in `indeterminate` and in the current note tags
    * This tag should be kept as checked
    * And as normal add all tags from  `selected`

Arthur-Milchior reviewed (commented):
Thanks a lot for all of those improvements. Splitting in smaller classes, adding tests, cleaning, commenting, that's really cool to see.

I am sorry because I still would like to ask a lot of changes. Concer

Concerning UniqueArrayList, I'd consider that https://docs.oracle.com/javase/7/docs/api/java/util/LinkedHashSet.html would probably be more standards.
Sadly, that's not perfect either, since you can't directly tell it to ignore string case in equality. However, I find that creating a class as described in https://stackoverflow.com/posts/24558633/revisions , that encapsulate the string with an hash function and equality that ignore case would lead to smaller and simpler code.

I'd suggest you to come ask us whether we know something that does the same job and already exists. It's not easy when to come and ask this kind of question, especially since you found a way to do the job and it really works, so I understand if you believe that my advice can't really be followed.

I'd like you to add "NF:" to start commits which are not functionals. So that I know that I must check that nothing changes in those commits. 
Speaking of which, I suspect that the first commit is NF, but I'm not really sure. It's a big commit. I would really love you to split it in part.  For example one commit just for putting `mType` into an enum (that's really cool to do that), then one to change initialization. Finally, maybe one which uses `requireActivity` and explain why. Right now, I feel like it's hard to check this first commit.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -39,18 +40,42 @@
+    @NonNull
+    private DialogType mType = DialogType.NONE;
+    @NonNull
+    private TreeSet<String> mCurrentTags = new TreeSet<>();
```
I did add initialization, to be able to annotate with @NonNull

TarekkMA replied:
But the mCurrentTags is removed at later commits.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:
```diff
@@ -0,0 +1,187 @@
+ * <p>
+ * The implementation of this class uses {@link ArrayList} as a base, to enable easy sorting and indexing.
+ */
+public class UniqueArrayList<E> extends ArrayList<E> {
```
I did create a new class, to be able to do the following. Keep order, have a comparator that will be used to check equality, and be able to sort the array.

I tried to use [SetUniqueList](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/list/SetUniqueList.html) but sorting was an issue.

In LinkedHashSet also the same issue but with being unable to index the element. It would require looping over all items, and this will not be suitable for RecyclerView where access to elements via index will be needed alot.

The best thing is to combine a set with a list and this what UniqueArrayList does, just delegate all methods to the Set and the underlying ArrayList.

Arthur-Milchior replied:
Makes perfect sens now. Can you please add such an indication in the comment, I would expect other people to have similar question.

When do you sort the list/set ? I can't find this information in the list of modification. I do wonder whether it must really be the same data structure that deal with all at once.

TarekkMA replied:
You can find sorting in `TagsList#sort`.

TarekkMA replied:
I will write a comment on this PR explaining my thought process of why I did create this class

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -47,6 +47,7 @@ public TagsList(List<String> allTags, List<String> checked) {
         mChecked = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
         mChecked.addAll(checked);
         mAllTags = UniqueArrayList.fromCollection(allTags, String.CASE_INSENSITIVE_ORDER);
+        mAllTags.addAll(mChecked);
```
I did try to keep the behavior the same, and not change the functionality after refactoring.

I don't know why this step was done on the legacy code. I did it on its own commit, so we can drop it if it's indeed useless (and it's from looking at where the dialog is being used).

The only why I can find it to be of any use if the caller sent elements in `checked` that is not in `allTags`

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -39,18 +40,42 @@
+         *
+         * The reason why it's present is to initialize {@link #mType} with an initial value
+         */
+        NONE
```
Issue is I want it to be @NonNull, I can drop NONE and pick the first element as a placeholder.

or create `requireDialogType()` which will return it as NonNull, throwing otherwise

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -68,6 +68,7 @@ public TagsList(List<String> allTags, List<String> checked) {
             mChecked = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
             mChecked.addAll(checked);
             mAllTags = UniqueArrayList.fromCollection(allTags);
+            mAllTags.addAll(mChecked);
```
Right sorry that's a bad rebase. I did add it in another commit that does explain why I added this. Will remove it from here

Arthur-Milchior commented:
On second thought, I feel like this PR will be very long to review because  a lot of things are happening at the same time.

If you're okay with that, I'd like you to do multiple PR first:
* one which deal only with cleaning the code, splitting it into parts, doing absolutely no functional change. Hopefully, it will be easy to reivew and can be merged quickly. Plus AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsListTest.java if it's possible
* One in which you introduce the data structure and their tests. You'll add an explanation stating that this PR will be useful for #8322 and that the goal is to make PR simples to understand. That mean during a few hours or days we'll have a unused data structure in our codebase, that's really not a problem. You already have merged code, this PR is here and show that you know your goal, so I believe it can be accepted easily

And when both PR are merged then we'll come back to current PR, which will have became smaller and easy to review.

I don't know whether it's possible and make sens. Maybe it would be nice if you could add an interface for your data structure. So that we see what are the functions that are actually used by your code. Once I see what function you expect to use, it will be easier to think about whether the data structure can be more efficient. Also, it would be nice to document how many tag this will contain. If you expect to work with all tags from the collection, it must be efficient, since some people have a lot and a lot of tags. However, if it only deals with the tags of a single note, I'd be fine with an easier code that is less efficient, at least in a first time

TarekkMA commented:
Thank you for your feedback, no worries I anticipated some changes.
<hr>

### My thought process when I created  `UniqueArrayList` 

The `UniqueArrayList<E>` class is created to be able to do the following:
*  **Maintain the order of the items**
*  **Indexable**: getting index of an element and vice versa
   * Needed in recycler view where it requires access to element via index a lot while scrolling.
*  **Sortable**: need a way to sort the array
    * Needed when adding tags or loading the Dialog, where checked tags must be shown first
*   **Unique**: maintaining a unique list of items using comparators
    * making tags unique with `String.CASE_INSENSITIVE_ORDER`
    
👎  `ArrayList<E>`: not Unique
👎   `TreeSet<E>`: doesn't maintain the order and no indexing and not sortable
👎  `LinkedHashSet<E>`: not indexable and cannot be sorted and not uniqe with custom comparator
👎  [`SetUniqueList<E>`](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/list/SetUniqueList.html): almost perfect but not sortable and no comparator support.

So I decided to combine both `ArrayList<E>` while keeping its underlying implementation with an internal ``Set<E>`` to check if an element is already in the list or not efficiently.
<hr>

I will try to break up the first commit into smaller commits and add NF and follow the comments you left on the code

Thank you again for the feadback, and if there is anything else I missed please let me know.

TarekkMA commented:
Okay so if I understood you correctly this will be 3 PRs

1. Cleaning PR, splitting the code
    * note that `TagsListTest` is part of the data structures that I introduced, the old code keeps 1 `ArrayList` and 1 `TreeSet` and keeping them in sync in the `TagsDialog`.
    * `TagsList` aim to remove that from  `TagsDialog` in return making it testable
    * But `TagsList` depend on the new `UniqueArrayList<E>`
2. Data Structures PR, along with their tests
    * `TagsList`
    * `UniqueArrayList<E>`
3. Utilize DS to refactor  `TagsDialog`
4. Fix #8322

<hr>

I don't know if this what you meant by interfaces.

## DS Interfaces

<details>
<summary>TagsList</summary>

```java
/**
 * A container class that keeps track of tags and their status
 */
public interface ITagsList extends Iterable<String> {
    /**
     * Return true if a tag is checked given its index in the list
     *
     * @param index index of the tag to check
     * @return whether the tag is checked or not
     * @throws IndexOutOfBoundsException if the index is out of range
     *                                   (<tt>index &lt; 0 || index &gt;= size()</tt>)
     */
    boolean isChecked(int index);

    /**
     * Return true if a tag is checked
     *
     * @param tag the tag to check
     * @return whether the tag is checked or not
     */
    boolean isChecked(String tag);

    /**
     * Adds an tag to the list if it is not already present.
     *
     * @param tag  the tag to add
     * @return true if tag was added (new tag)
     */
    boolean add(String tag);

    /**
     * Mark a tag as checked tag
     *
     * @param tag the tag to be checked
     * @return true if the tag changed its check status
     *         false if the tag was already checked or not in the list
     */
    boolean check(String tag);

    /**
     * Mark a tag as unchecked tag
     *
     * @param tag the tag to be checked
     * @return true if the tag changed its check status
     *         false if the tag was already unchecked or not in the list
     */
    boolean uncheck(String tag);

    /**
     * Toggle the status if all tags,
     * if all tags are checked, then uncheck them
     * otherwise uncheck all tags
     *
     * @return true if this tag list changed as a result of the call
     */
    boolean toggleAllCheckedStatuses();

    /**
     * @return Number of tags in the list
     */
    int size();

    /**
     * Returns the tag at the specified position in this list.
     *
     * @param index index of the tag to return
     * @return the tag at the specified position in this list
     * @throws IndexOutOfBoundsException if the index is out of range
     *                                   (<tt>index &lt; 0 || index &gt;= size()</tt>)
     */
    @NonNull
    String get(int index);

    /**
     * @return true if there is no tags in the list
     */
    boolean isEmpty();

    /**
     * Return all checked tags as an List that maybe not in order
     */
    List<String> getCheckedTagList();

    /**
     * @return return all tags as an List
     */
    List<String> getAllTagList();

    /**
     * Sort the tag list alphabetically ignoring the case, with priority for checked tags
     */
    void sort();
}
```

</details>

<details>
<summary>UniqueArrayList</summary>

```java
/**
 * A decorator over {@link ArrayList} to make it unique, to ensure that all inserted elements is unique while maintaining the order
 * <p>
 * The implementation of this class uses {@link ArrayList} as a base, to enable easy sorting and indexing.
 */
public interface IUniqueArrayList<E> extends List<E>, RandomAccess {
    /**
     * Construct a {@link UniqueArrayList} from a {@link Collection} after removing any duplicates
     *
     * @param c the source collection
     * @param comparator the comparator used to check element equality
     */
    @SuppressWarnings("SortedCollectionWithNonComparableKeys")
    static <E> UniqueArrayList<E> fromCollection(@NonNull Collection<? extends E> c, @Nullable Comparator<? super E> comparator) {
        ///redacted
    }

    /**
     * Construct a {@link UniqueArrayList} from a {@link Collection} after removing any duplicates
     *
     * @param c the source collection
     * @param comparator the comparator used to check element equality
     */
    static <E> UniqueArrayList<E> fromCollection(@NonNull Collection<? extends E> c) {
        return fromCollection(c, null);
    }

    /**
     * Appends the specified element to the end of this list if it is unique.
     *
     * @param e element to be appended to this list
     * @return true if the element is added
     */
    @Override
    boolean add(E e);

    /**
     * Inserts the specified element at the specified position in this
     * list if it is unique. Shifts the element currently at that position (if any) and
     * any subsequent elements to the right (adds one to their indices).
     *
     * @param index index at which the specified element is to be inserted
     * @param element element to be inserted
     * @deprecated use {@link #addAt(int, Object)} instead
     */
    @Override
    @Deprecated
    void add(int index, E element);

    /**
     * Inserts the specified element at the specified position in this
     * list if it is unique. Shifts the element currently at that position (if any) and
     * any subsequent elements to the right (adds one to their indices).
     *
     * @param index index at which the specified element is to be inserted
     * @param element element to be inserted
     * @return true if the element is added
     */
    boolean addAt(int index, E element);

    /**
     * Replaces the element at the specified position in this list with
     * the specified element if the specified is unique.
     *
     * @param index index of the element to replace
     * @param element element to be stored at the specified position
     * @return if replaced, the element previously at the specified position
     * if element is not unique null.
     * @see #setAt(int, E)
     */
    @Override
    @Nullable
    E set(int index, E element);

    /**
     * Replaces the element at the specified position in this list with
     * the specified element if the specified is unique.
     *
     * @param index index of the element to replace
     * @param element element to be stored at the specified position
     * @return true if the element is replaced
     */
    boolean setAt(int index, E element);

    /**
     * Adds all unique elements to the list
     *
     * @param c collection containing elements to be added to this collection
     * @return true if this collection changed as a result of the call
     */
    @Override
    boolean addAll(@NonNull Collection<? extends E> c);

    /**
     * Adds all unique elements to the list at a specified index
     *
     * @param c collection containing elements to be added to this collection
     * @return true if this collection changed as a result of the call
     */
    @Override
    boolean addAll(int index, @NonNull Collection<? extends E> c);

    @Override
    void clear();

    @Override
    E remove(int index);

    @Override
    boolean remove(@Nullable Object o);

    @Override
    boolean removeAll(@NonNull Collection<?> c);

    /**
     * @return true if the element is found in the list
     */
    @Override
    boolean contains(@Nullable Object o);

    @Override
    @Deprecated
    boolean removeIf(@NonNull Predicate<? super E> filter);

    @Override
    @Deprecated
    void replaceAll(@NonNull UnaryOperator<E> operator);
}
```
</details>

TarekkMA commented:
@Arthur-Milchior after I give thought about the efficiency, I think we can do better than `TreeSet<String>`. We can use `HashSet<String>` with the modification you linked, the lookup/adding/removing time will be much faster `O(1)` instead of `O(log (n))`'s of `TreeSet`.

If we are going with this I will need to add a new class that implements `Set<String>` and proxies calls to an internal `HashSet<String>` after wrapping  string in a wrapper `CaseInsensitiveStringSet`, the class name might be `CaseInsensitiveStringSet`.

I think the `CaseInsensitiveStringSet` in many parts of the code where handling of tags is required.

<details>

<summary>Code Sample</summary>

```java
public class CaseInsensitiveString implements CharSequence {

    public CaseInsensitiveString(String src){...}

    @Override
    public int hashCode() {...}

    @Override
    public boolean equals(Object obj) {...}
}



class CaseInsensitiveStringSet implements Set<String> {
    
    final HashSet<CaseInsensitiveString> mHashSet = new HashSet<>();

    ....
    
    @Override
    public boolean contains(@Nullable Object o) {
        return mHashSet.contains(new CaseInsensitiveString((String) o));
    }


    @Override
    public boolean add(String s) {
        return mHashSet.add(new CaseInsensitiveString(s));
    }


    @Override
    public boolean remove(@Nullable Object o) {
        return mHashSet.add(new CaseInsensitiveString((String) o));
    }
    
    .....
}
```
</details>

Arthur-Milchior commented:
> ## DS Interfaces
> I don't know if this what you meant by interfaces.

`IUniqueArrayList` is indeed similar to what I meant. I did not expect that so many methods would be needed, but I trust you.

I was not asking for ITagsList. The difference for me is that I can imagine that `UniqueArrayList` can be kept quite abstract, while the Tag List is really related to the tags themselves. For example they hard code the notion of checked values. 

I'm fine with not worrying about having the optimal implementation yet. A simple implementation is really nice. At best, it would be cool to check how much time it takes on a collection will, let's say, 10000 tags. If it runs quickly, then it'll be fine. No need to spend time optimizing something that is not a bottle neck.

TarekkMA commented:
I needed `UniqueArrayList` to be compatible with the Collection/List interfaces, so I can use it with anything that the standard library provides, like `Collection.sort(list, comparator)`. So I chose `ArrayList` as a base, and every method where data in the list will be changed I had to sync the internal `Set`.

I did only add `addAt` and `setAt` as a varient to `add` and `set`, where I can return a boolean to represent if a change did occur. 

I will try to benchmark `UniqueArrayList` against some of other collections.

TarekkMA commented:
@Arthur-Milchior, I think results look okay except the remove one, and I think I know why. (will update after I fix it).
```
benchmark:           473 ns ExampleBenchmark.UniqueArrayList_contains
benchmark:        56,510 ns ExampleBenchmark.ArrayList_contains
benchmark:           459 ns ExampleBenchmark.TreeSet_contains
benchmark:           295 ns ExampleBenchmark.LinkedHashSet_contains

benchmark:         1,386 ns ExampleBenchmark.UniqueArrayList_add
benchmark:         1,562 ns ExampleBenchmark.ArrayList_add
benchmark:         1,406 ns ExampleBenchmark.TreeSet_add
benchmark:         1,041 ns ExampleBenchmark.LinkedHashSet_add

benchmark:        99,869 ns ExampleBenchmark.UniqueArrayList_remove
benchmark:        56,705 ns ExampleBenchmark.ArrayList_remove
benchmark:           470 ns ExampleBenchmark.TreeSet_remove
benchmark:           293 ns ExampleBenchmark.LinkedHashSet_remove
```

### Test Setup
Was done using `androidx.benchmark:benchmark-junit4` on Huawei P30 phone.

### Code Sample
```java
    @Test
    public void UniqueArrayList_add() {
        final BenchmarkState state = mBenchmarkRule.getState();
        final Collection<String> collection = UniqueArrayList.fromCollection(new ArrayList<>());
        while (state.keepRunning()) {
            collection.add(RandomStringUtils.random(10));
        }
    }
    
    @Test
    public void ArrayList_remove() {
        final BenchmarkState state = mBenchmarkRule.getState();
        final Collection<String> collection = new ArrayList<>();
        fillWithRandomStrings(collection,10_000);
        while (state.keepRunning()) {
            collection.remove(RandomStringUtils.random(10));
        }
    }
    
    
    @Test
    public void TreeSet_contains() {
        final BenchmarkState state = mBenchmarkRule.getState();
        final Collection<String> collection = new TreeSet<>();
        fillWithRandomStrings(collection,10_000);
        while (state.keepRunning()) {
            collection.contains(RandomStringUtils.random(10));
        }
    }
```

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,193 @@
+     * @return true if the tag changed its check status
+     *         false if the tag was already unchecked or not in the list
+     */
+    public boolean uncheck(String tag) {
```
It seems counter intuitive to me. Usually, if setter return something, it's the previous value. So I'd assume it return the same false if the value was already unchecked.

Implementer choice

TarekkMA replied:
Here I looked to the `Set` interface for ideas. And it does return boolean to indicate if a change did happen or not.

Will try to split this PR into the 4 I listed above, hope it will be organized.

Arthur-Milchior replied:
Thanks for the explanation. Fine then

Arthur-Milchior commented:
If there is a delay of only a tenth of second when a user ask to remove a tag, that would be fine by me. I mean, it's not like they can remove 10 tags by second. Not even one by second.

But, I insist, I would really like to split discussion in multiple PR. If you could just use the code you wrote, create a PR for the data structure and we can keep the discussion here, separated from other concerns, that would be great

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java:
```diff
@@ -214,37 +235,36 @@ public boolean onQueryTextChange(String newText) {
+     * A wrapper function around dialog.getInputEditText() to get non null {@link EditText}
+     */
+    @NonNull
+    private EditText requireDialogInputEditText(MaterialDialog dialog){
```
```suggestion
    private EditText requireDialogInputEditText(@NonNull MaterialDialog dialog) {
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,193 @@
+ */
+public class TagsList implements Iterable<String> {
+    @NonNull
+    final private TreeSet<String> mChecked;
```
```suggestion
    private final TreeSet<String> mChecked;
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,193 @@
+    @NonNull
+    final private TreeSet<String> mChecked;
+    @NonNull
+    final private UniqueArrayList<String> mAllTags;
```
```suggestion
    private final UniqueArrayList<String> mAllTags;
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,193 @@
+
+
+    /**
+     * Adds an tag to the list if it is not already present.
```
```suggestion
     * Adds a tag to the list if it is not already present.
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,193 @@
+     *         false if the tag was already checked or not in the list
+     */
+    public boolean check(String tag) {
+        if (!mAllTags.contains(tag)) {
```
Can we state case-sensitivity in the method docs? (also a few later on)

TarekkMA replied:
I think it's better to state that on the entire class, since this is how the `TagsList` class works

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,193 @@
+     * @return true if this tag list changed as a result of the call
+     */
+    public boolean toggleAllCheckedStatuses(){
+        if(mAllTags.size() == mChecked.size()){
```
```suggestion
        if (mAllTags.size() == mChecked.size()) {
```

TarekkMA replied:
fixed

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,193 @@
+        if(mAllTags.size() == mChecked.size()){
+            mChecked.clear();
+            return true;
+        }else{
```
```suggestion
        } else {
```

TarekkMA replied:
fixed

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/utils/UniqueArrayList.java:
```diff
@@ -0,0 +1,280 @@
+ * <p>
+ * The implementation of this class uses {@link ArrayList} as a base, to enable easy sorting and indexing.
+ */
+public class UniqueArrayList<E> extends ArrayList<E> {
```
A problem with extending `ArrayList`, rather than encapsulating `ArrayList` and implementing `Collection` would be that any new methods added to `ArrayList` wouldn't be overridden, and the `Set` would be out-of-sync with the collection.

TarekkMA replied:
composition over inheritance. Alright then.

david-allison-1 replied:
Shame it makes the code more verbose, but yeah

david-allison-1 reviewed (commented):
Fantastic amount of testing! Looks good to me, minor number of nitpicks

Could you remove the `Fixes #XXX` from the description, we don't want these closed unless the change actually affects CardBrowser

Could you also change the name of the PR to explain that it's mostly prep work for the change?

TarekkMA commented:
> But, I insist, I would really like to split discussion in multiple PR. If you could just use the code you wrote, create a PR for the data structure and we can keep the discussion here, separated from other concerns, that would be great

@Arthur-Milchior  okay I will.

TarekkMA commented:
@Arthur-Milchior did create a PR for the DS

Arthur-Milchior commented:
I put it as draft. So that it's clear when looking at the list it waits for other PR to be merged and can't be considered imediately. Feel free to un-draft it as soon as you have rebased and we should review

david-allison-1 commented:
@TarekkMA I believe this can be worked on now, please let me know if I'm mistaken

TarekkMA commented:
> @TarekkMA I believe this can be worked on now, please let me know if I'm mistaken

You are right, there are no blockers for this PR. will do refactorings in the PR and will follow that with the multi tag edit feature.

But I won't be available until tomorrow night.

david-allison-1 commented:
No rush, just making sure I hadn't missed anything

TarekkMA commented:
I have synced the branch with master and solved all merge conflicts.

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsListTest.java:
```diff
@@ -0,0 +1,214 @@
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertEquals;
+
+@SuppressWarnings("ArraysAsListWithZeroOrOneArgument")
```
Is there a reason for not using singletonlist?

TarekkMA replied:
No, just for symmetry with other code 😄

david-allison-1 replied:
Reasonable - Leave it in with a comment

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,194 @@
+        if (mAllTags.size() == mChecked.size()) {
+            mChecked.clear();
+            return true;
+        } else {
```
Optional: no need for the `else`

TarekkMA replied:
right

david-allison-1 reviewed (changes requested):
Looks good! One main comment:

* Many of the assertions will not provide a good message if they fail, since you have the best mental context of the change now, could you add in messages where a failure would be non-obvious

TarekkMA commented:
> * Many of the assertions will not provide a good message if they fail, since you have the best mental context of the change now, could you add in messages where a failure would be non-obvious

if you mean the tests, I will add helpful messages to them.

TarekkMA commented:
@david-allison-1 did add some messages, and other I think the test name should be enough.

david-allison-1 reviewed (approved):
LGTM, thanks

david-allison-1 commented:
@Arthur-Milchior, all yours

Arthur-Milchior commented:
I'm having a hard time following the first commit. I tried to atomize it as much as possible in #8479 and would like your help to understand some part of it

TarekkMA commented:
@Arthur-Milchior based on the discussion we had on #8479 I did the following:
- rebased your PR
- edited the  commit message of `NF: Simplify adjustToolBar` to give more context on the toggle all function
- split the last commit in your PR to 5 commits
- some small fixes to commits in this PR.

mikehardy reviewed (approved):
This LGTM - it seemed like @Arthur-Milchior may have had a pending comment but it also may be handled, I'm going to mark it with the rare combo of "Pending Merge" and "Waiting for Feedback"

In my opinion "a few days" (let's say 3 to be concrete) is enough for any further feedback if there is any, otherwise I or anyone else that can merge should feel free to merge? Seems reasonable as I type it...

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsList.java:
```diff
@@ -0,0 +1,194 @@
+/*
```
I believe the commit "Move TagsList to its own class" should also delete the code from the current `TagsList` subclass. This seems especially important because it makes really simple to check, just by looking at the commit message, that the change is correct that it comes from a copy/paste or from an automated refactoring.

The risk is that, if at some point, you make a correction in the previous Commits, then this correction is not reflected here. (An easy way to decrease this risk is simply to first merge all commits before the moving commits, and then move the moving commits in a distinct PR, when we are sure that the original class won't be changed again)

TarekkMA replied:
Oh I think after lots of rebasing and solving conflicts that happened. 

Should I drop the move commits from this PR and do it in another one?

```
Move TagsArrayAdapter to its own class
Move TagsList to its own class
```

Arthur-Milchior replied:
> Oh I think after lots of rebasing and solving conflicts that happened. 

I don't understand


and yes, this is what I suggested. Moving in a distinct PR will really help, as it ensure that you don't have to deal with conflict each time you correct a commit

TarekkMA replied:
I meant that whenever I chnge something in previous commits, there will be a conflict with the move commit. 

I guess I did forget to handle it correctly

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsArrayAdapter.java:
```diff
@@ -0,0 +1,117 @@
+package com.ichi2.anki.dialogs.tags;
```
Note for the person who merge. Pay close attention that the commit "Move TagsArrayAdapter to its own class" actually consists simply in copy/paste and don't introduce change.

I won't review immediately since I asked for some change in previous commits

TarekkMA replied:
I did remove all move commits, another PR for only move commits as Arthur suggested will be trivial.

Arthur-Milchior reviewed (commented):
Thank you very much for this huge work. 
It still took a lot of time to review, but at least I feel like I could be confident that I understand what I review, and I really appreciate your help in it.

I have still a little bit of remarks.
Then, once everything is actually corrected as we want, I'll want a person to look closely and check that the commits that move code only move it and don't introduce change. An accident arrives far too easily in this kind of PR (I was the faulty one sometime)

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -507,33 +520,25 @@ public Filter getFilter() {
             @Override
             protected FilterResults performFiltering(CharSequence constraint) {
-                mFilteredTags.clear();
                 if (constraint.length() == 0) {
```
`CharSequence` don't seem to have `isEmpty`

Arthur-Milchior replied:
My bad.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -479,7 +481,7 @@ public void sortData() {
                 ctv.toggle();
                 String tag = ctv.getText().toString();
-                if (ctv.isChecked()) {
+                if (ctv.isChecked() && !mTags.isChecked(tag)) {
```
I didn't add it, I simply did continue to refactor it until the commit 
`52211898 Remove extra check from TagsArrayAdapter` where I did finally removed it

TarekkMA replied:
oh sorry, I thought you were talking about an earlier commit.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -71,6 +71,32 @@ public boolean isChecked(int index) {
+         *         false if the tag was already checked or not in the list
+         */
+        public boolean check(String tag) {
+            if (!mAllTags.contains(tag)) {
```
AFAIK the current code doesn't check a tag that is not in all tags, so you are right I did add this extra if.

I did add it as a guard so the checked set doesn't go out of sync from all tags.

Should I split it into two commits ?

Arthur-Milchior replied:
If you feel like the guard should be kept, then yes.

This do not answer my question. Is it something we actually need to guard against ?
As far as I'm concerned, it's perfectly okay to assume that `tag` belong to `mAllTags` and to require it in the javadoc instead of spending time checking for the tag in the set.
Similarly, You have not stated that `tag` should be `@NonNull` (you probably should), that's not a reason which justify to check whether the String is null even if it theoretically could be called with a null value

TarekkMA replied:
> This do not answer my question. Is it something we actually need to guard against ?

I think so, because we do return the list of all checked tags to the caller, via `TagsDialogListener#onSelectedTags(List<String> `**`selectedTags`**`, int option)`

I made it so it's a future proof. so this little detail doesn't introduce a bug later on.

Arthur-Milchior replied:
Okay. Then please keep it. Just use a different commit so that it's really clear why this change is introduced.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -294,18 +294,12 @@ public boolean onQueryTextChange(String newText) {
-            if (mTags.mCurrentTags.containsAll(mTagsArrayAdapter.mTagsList)) {
-                mTags.mCurrentTags.removeAll(mTagsArrayAdapter.mTagsList);
+            if (mTags.mAllTags.size() == mTags.mCurrentTags.size()) {
+                mTags.mCurrentTags.clear();
```
thank you

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -121,6 +121,19 @@ public boolean toggleAllCheckedStatuses() {
+                boolean lhs_checked = isChecked(lhs);
+                boolean rhs_checked = isChecked(rhs);
+                //priority for checked items.
+                return lhs_checked == rhs_checked ? lhs.compareToIgnoreCase(rhs) : lhs_checked ? -1 : 1;
```
How about this? I will add it in another commit after this one.
```java
        /**
         * Sort the tag list alphabetically ignoring the case, with priority for checked tags
         */
        public void sort() {
            Collections.sort(mAllTags, (lhs, rhs) -> {
                boolean lhsChecked = isChecked(lhs);
                boolean rhsChecked = isChecked(rhs);
                
                if (lhsChecked == rhsChecked) {
                    return lhs.compareToIgnoreCase(rhs);
                } else {
                    // checked tags must appear first
                    return lhsChecked ? -1 : 1;
                }
            });
        }
```

Arthur-Milchior replied:
I've a small preference for exchanging if and else. Indeed, logically it seems easier to me in english to write:
> If only one tag is checked, it should appear first, otherwise we use alphabetical order

And this exchange would reflect the usual english sentence

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -479,9 +479,9 @@ public void sortData() {
                 ctv.toggle();
                 String tag = ctv.getText().toString();
-                if (ctv.isChecked() && !mTags.isChecked(tag)) {
+                if (ctv.isChecked()) {
```
I could write the commit message as follows:

```
- change to the check status should be driven from the checkbox status
- in the case that the tag is already checked `mTags.check(tag)` won't change anything.
```

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -46,8 +46,22 @@
+         *
+         * @param allTags A list of all tags possible
+         *                any duplicates will be ignored
+         * @param currentTags A list of checked tags
```
Can you explain the last sentence?
> it is okay to just copy-paste the description in both places and wait for the commits that change the constructor to change the related comments

Arthur-Milchior replied:
I asked you to add javadoc to explain what is `mAllTags` and `mCurrentTags`

You can use this same javadoc to document the constructors' parameters in the first commit introducing a non-trivial constructor.

Since you change the constructor twice, it would be nice to change the javadoc twice to reflect the changes. So that each commit contain an accurate javadoc

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -507,33 +520,25 @@ public Filter getFilter() {
-                mFilteredTags.clear();
                 if (constraint.length() == 0) {
-                    mFilteredTags.addAll(mTags.mAllTags);
+                    mFilteredList = null;
```
The adapter class had the benefit of accessing the `TagsDialog#mTags`, but in the next commit `decouple the adapter from `TagsDialog`` I made the adapter uncoupled from the dialog.

So instead of relying on `TagsDialog#mTags` to get all tags unfiltered, the adapter have to store a reference/copy to the original data inside it.

TarekkMA replied:
do you suggest having one List instance to use and another one to hold a reference to the original one?

so we use `mFilteredList` in all things? and have `mTagsList` just to check tag status, and hold reference to all tags?

Arthur-Milchior replied:
Please don't use backquote for commit message, as it's not code (and it breaked in your text).


I believe you understood my suggestion right. 
You may disagree if you think it's a bad idea. I would simply like to see javadoc explaining in details the assumptions you made (e.g. "`mFilteredList` is the subset of `mTagsList` satisfying the user's search") and a quick explanation in the commit message

TarekkMA commented:
@Arthur-Milchior 
* Dropped move commits (will be done in another PR following this)
* Fixed all issues you pointed.

please let me know if I did miss anything.

Thank you.

mikehardy commented:
Just a note as I am scanning "Needs Second Approval" label for merges - I think this one is waiting on Arthur (and I re-requested review from him) but because of the GSoC admin/mentor load he likely just does not have time at the moment. But for sure there will be in the future, at some point :-). Thanks for your patience!

TarekkMA commented:
I do fully understand that, currently there are almoy 100 Open PRs 😅. So I hope you all the best and good luck.

Thanks

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -103,6 +103,22 @@ public boolean check(String tag) {
+        /**
+         * Toggle the status if all tags,
+         * if all tags are checked, then uncheck them
+         * otherwise uncheck all tags
```
otherwise check all tags

I believe you made a typo

TarekkMA replied:
fixed
```java
        /**
         * Toggle the status of all tags,
         * if all tags are checked, then uncheck them
         * otherwise check all tags
         *
         * @return true if this tag list changed as a result of the call
         */
```

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -299,14 +315,7 @@ public boolean onQueryTextChange(String newText) {
-                changed = mTags.mCurrentTags.addAll(mTags.mAllTags);
-            }
-            if (changed) {
+            if (mTags.toggleAllCheckedStatuses()) {
```
I'd add a variable to save `mTags.toggleAllCheckedStatuses()`

It's technically useless, but I think that it's not clear why there is an action in an `if`. A variable `setOfCheckedTagsChanged` would make the meaning clear

TarekkMA replied:
```java
            final boolean didChange = mTags.toggleAllCheckedStatuses();
            if (didChange) {
                mTagsArrayAdapter.notifyDataSetChanged();
            }
```

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -28,19 +23,208 @@
+
+        @NonNull
+        @Override
+        public Iterator<String> iterator() {
```
I'd appreciate if your javadoc, either here or on the class, can indicate that iteration is done on all tags, independently on whether they are checked

TarekkMA replied:
Right I will make things more clear

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -388,12 +401,7 @@ public  TagsArrayAdapter() {
         }
 
         public void sortData() {
-            Collections.sort(mTagsList, (lhs, rhs) -> {
```
Is this valid?
I want that all commits compile. In the commit, `mTagsList` is not yet an iterable or a collection, so I don't see how it could work

Arthur-Milchior replied:
This is the only thing that block me from merging. All other changes are small and could be corrected in a later PR

TarekkMA replied:
I can also change `mTagsList` to be a reference to `TagsList` class. I think I had it like this before, but reverted it for some reason.

Arthur-Milchior replied:
Sorry, it's not clear to me. Should I review now, or is it something you'll do later and I should wait?

TarekkMA replied:
Sorry, I wasn't clear, I will provide more information on what I meant by my comment.

This change is only relevant in "NF: move `sort` in `TagList`" commit. and I did add it there, in later commits I don't need to do so because the `TagsList` reference is used rather than a reference to a list.

```java
// "NF: move `sort` in `TagList`" commit
        public  TagsArrayAdapter() {
            mTagsList = new ArrayList<>(mTags.mAllTags);
            sortData();
        }

        public void sortData() {
            mTags.sort();
            mTagsList = new ArrayList<>(mTags.mAllTags);
        }
```
```java
//new commits     
        public TagsArrayAdapter(@NonNull TagsList tags) {
            mTags = tags;
            mFilteredList = new ArrayList<>(tags.mAllTags);
            sortData();
        }

        public void sortData() {
            mTags.sort();
        }
```

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -141,6 +141,14 @@ public boolean isEmpty() {
+        /**
+         * @return return a copy of checked tags
+         */
+        public List<String> getCheckedTagList() {
```
Do you know whether `onSelectedTags` modify the list? I would imagine that creating a copy is actually useless. Not really important in practice as I expect the list to be small

TarekkMA replied:
In all references to `onSelectedTags` the checked list is not changed at all.

I can return `Collections.unmodifiableList(mAllTags)`, that way we avoid copy and ensure no one can mutate the data, and if needed the caller should do a copy in his side

Arthur-Milchior replied:
I like the idea. Might be worth to check other occurrences of `new DataStructureType(containerToBeCopied)`. But clearly not in this PR

Arthur-Milchior replied:
let me know when it's done and I'll review the PR again then

TarekkMA replied:
This PR is done and ready for review

Arthur-Milchior replied:
FYI: I don't get notified when you put an emoticon. So if you want me to know that you have read it and accept, you need to post an actual text

TarekkMA replied:
Oh okay I will do that in the future

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -459,19 +459,31 @@ public void addTag(String tag) {
+        @NonNull
+        private List<String> mFilteredList;
+
+        public  TagsArrayAdapter(@NonNull TagsList tags) {
```
nit: a single space after `public`

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialog.java:
```diff
@@ -518,33 +530,25 @@ public Filter getFilter() {
 
         /* Custom Filter class - as seen in http://stackoverflow.com/a/29792313/1332026 */
         private class TagsFilter extends Filter {
```
Is this inner class still useful ? It seems to contain no state, so I don't see its point

TarekkMA replied:
The recycler view implements the [`Filterable`](https://developer.android.com/reference/android/widget/Filterable.html) interface which needs an instance to `Filter`.

I can refactor this so the filtering action is done via a method attached to the adapter, should I do that?

Arthur-Milchior replied:
Eventually, it seems to make sens. Clearly not in this PR, it's already quite big enough

Arthur-Milchior reviewed (commented):
First, thank you a lot for this hard work. Both the decoupling is great, and ALSO the commit history. It's really really easier to review this PR than the first PR I recalled reviewing from you not even two months ago.

As Mike told, things have been overwhelming, and clearly big PR were slowed down as smaller PR were easier to review. Even if everything was simple to review here, it still takes time to check 22 commits.
Most of them I would merge right now if it were possible. There remains only a few nits and questions. 

For future merger, of course, rebase, don't squash!

Arthur-Milchior commented:
My question above remain: should I start reviewing again now? Or should I wait for more changes to occur related to some discussions that are not marked as resolved?

TarekkMA commented:
It's ready for review

Arthur-Milchior commented:
Finally done!

Thank you again for all of this work. I know for some reason I asked for more than David and Mike, I fear that honestly, it was less easy for me to check the validity than them, but I must admit I'm still quite happy about the result you got and really prefer the current version.

And now, moving those classes to a new file will be really quick to proof read and merge, I believe it to be a huge plus.

mikehardy commented:
Actually @Arthur-Milchior it is more that when I saw you being as detailed as you were I stopped reviewing and backed away slowly ;-), relying almost entirely on you to bring this one home with @TarekkMA - and you guys did :-) :partying_face:
